cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/ProgramRunDispatcherTask.java

```diff
@@ -0,0 +1,152 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * Implementation of {@link RunnableTask} to execute Program-run operation in a system worker.
+ */
+public class ProgramRunDispatcherTask implements RunnableTask {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunDispatcherTask.class);
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(
+          new GsonBuilder())
+      .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+      .registerTypeAdapter(ApplicationClass.class, new ApplicationClassCodec())
+      .registerTypeAdapter(Requirements.class, new RequirementsCodec())
+      .registerTypeAdapter(RunId.class, new RunIds.RunIdCodec())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+      .create();
+
+  private final Injector injector;
+
+  @Inject
+  ProgramRunDispatcherTask(Injector injector) {
+    this.injector = injector;
+  }
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    try {
+      ProgramRunInfo programRunInfo = GSON.fromJson(context.getParam(), ProgramRunInfo.class);
+      DispatchTaskRunner taskRunner = injector.getInstance(DispatchTaskRunner.class);
+      ProgramController programController = taskRunner.dispatch(programRunInfo);
+      RemoteProgramRunDispatcherResponse response = null;
+      if (Objects.nonNull(programController)) {
+        response = new RemoteProgramRunDispatcherResponse(true);
+      }
+      String result = GSON.toJson(response);
+      context.writeResult(result.getBytes(StandardCharsets.UTF_8));
+    } catch (Exception e) {
+      LOG.error("Exception: ", e);
+    }
+  }
+
+  private static class DispatchTaskRunner {
+
+    private final CConfiguration cConf;
+    private final ProgramRunnerFactory programRunnerFactory;
+    private final ConfiguratorFactory configuratorFactory;
+    private final Impersonator impersonator;
+    private final ArtifactRepository artifactRepository;
+    private ProgramRunnerFactory remoteProgramRunnerFactory;
+    private InetAddress host;
+
+    @Inject
+    DispatchTaskRunner(CConfiguration cConf,
+        ProgramRunnerFactory programRunnerFactory,
+        ConfiguratorFactory configuratorFactory,
+        Impersonator impersonator,
+        @Named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO) ArtifactRepository artifactRepository) {
+      this.cConf = cConf;
+      this.programRunnerFactory = programRunnerFactory;
+      this.configuratorFactory = configuratorFactory;
+      this.impersonator = impersonator;
+      this.artifactRepository = artifactRepository;
+    }
+
+    /**
+     * Optional guice injection for the {@link ProgramRunnerFactory} used for remote execution. It
+     * is optional because in unit-test we don't have need for that.
```

Comment:
similar question to above, which unit tests are using this? Can those tests bind these to sensible alternatives instead?

Reply:
Changed.